### DISCUSSION TITLE
Using the app in receive only mode

### DISCUSSION THREAD

**Discussion by KonstantinBYvitebsk** [stasel/WebRTC-iOS#98] [General]
Hi, there.
We trying to use the app  in receive video only mode.
But now we are stumbled in that we  need to send our offer and receive  answer. 
I would love it if you could  describe how  to skip the stages sending offer and receiving answer to just show video from other side.
We  do not need  to send our video or audio.

Thank you.

**Comment by stasel** [comment]
Hi, in order to establish a WebRTC connection you must exchange offer/answer between the peers.
It sounds like what you want to achieve can be done with HTTP stream (or similar technologies) - I suggest investigating those technologies

**Comment by KonstantinBYvitebsk** [comment]
Hi, stasel. Thank you for your reply.
I know about the offer/answer stage. And understand it.
I am trying to connect your application with device working by gstreamer webRTC plugin.
Today I found some mistakes at my logic, so I am trying to fix them now.
May be this topic should  be closed.

**Comment by KonstantinBYvitebsk** [comment]
Hi there.
Hi @stasel 

It seems to me I've deleted  all code related to sending video.
But I still have no video on receive side. What wrong? It works with browsers but not with this application. 
Here is my negotiations:
Negotiation offer created (from  device side):
```
v=0
o=- 2929954850166907103 0 IN IP4 0.0.0.0
s=-
t=0 0
a=ice-options:trickle
a=msid-semantic:WMS webrtcbin
m=video 9 UDP/TLS/RTP/SAVPF 96
c=IN IP4 0.0.0.0
a=setup:actpass
a=ice-ufrag:jMEgNzWL5p907pa7NfPvc/Op1rAuPiBg
a=ice-pwd:rA7hwOFj8KFEW3eeXgZV+Tca6JForPhs
a=sendonly
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 H264/90000
a=rtcp-fb:96 nack pli
a=framerate:15
a=fmtp:96 packetization-mode=1;profile-level-id=42401e;sprop-parameter-sets=Z0JAHqaAoC/5UA==,aM48gA==
a=ssrc:1353650531 msid:user2213278508@host-73f07fe9 webrtctransceiver0
a=ssrc:1353650531 cname:user2213278508@host-73f07fe9
a=mid:video0
a=fingerprint:sha-256 4F:1E:96:93:AD:3C:BD:72:0A:8D:F0:E3:F6:1B:8A:17:BA:F1:03:41:0D:4C:70:D6:D8:DA:50:88:58:B8:6F:51

```
 Answer (from iOs):
```
v=0
o=- 4605800069827021447 2 IN IP4 127.0.0.1
s=-
t=0 0
a=msid-semantic: WMS
m=video 9 UDP/TLS/RTP/SAVPF 96
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:ZUzR
a=ice-pwd:hsnDH2eVFxVD/4N9rXihjIs7
a=ice-options:trickle renomination
a=fingerprint:sha-256 F6:8C:D7:0B:50:A5:DC:A5:E1:5C:45:80:AE:59:BA:6E:7E:BC:78:46:68:23:2A:C0:12:57:5A:0A:77:2A:BD:98
a=setup:active
a=mid:video0
a=recvonly
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 H264/90000
a=rtcp-fb:96 nack pli
a=fmtp:96 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01e

```
Candidates (sent from device to iOs):
```
candidate:12 2 TCP 1010828286 192.168.15.111 51165 typ host tcptype passive
candidate:11 2 TCP 1015022590 192.168.15.111 9 typ host tcptype active
candidate:10 2 UDP 2013266427 192.168.15.111 36799 typ host
candidate:12 1 TCP 1010828287 192.168.15.111 41271 typ host tcptype passive
candidate:11 1 TCP 1015022591 192.168.15.111 9 typ host tcptype active
candidate:10 1 UDP 2013266428 192.168.15.111 49624 typ host
candidate:9 1 TCP 1010828031 192.168.11.100 51679 typ host tcptype passive
candidate:8 1 TCP 1015022335 192.168.11.100 9 typ host tcptype active
candidate:7 1 UDP 2013266429 192.168.11.100 55105 typ host

```  

Candidates (sent from iOs to my Device):
```
candidate:653531393 1 udp 2122260223 192.168.15.106 52142 typ host generation 0 ufrag ZUzR network-id 1 network-cost 10
candidate:3586273016 1 udp 2122197247 2a02:bf0:322c:f47c:a9e5:b45:b860:10e0 58368 typ host generation 0 ufrag ZUzR network-id 5 network-cost 900
candidate:1659853702 1 udp 2122129151 10.187.242.102 57484 typ host generation 0 ufrag ZUzR network-id 4 network-cost 900
candidate:1752688113 1 tcp 1518280447 192.168.15.106 55881 typ host tcptype passive generation 0 ufrag ZUzR network-id 1 network-cost 10
candidate:745318262 1 tcp 1518149375 10.187.242.102 55883 typ host tcptype passive generation 0 ufrag ZUzR network-id 4 network-cost 900
 candidate:3933808274 1 udp 1686052607 46.53.180.26 5664 typ srflx raddr 192.168.15.106 rport 52142 generation 0 ufrag ZUzR network-id 1 network-cost 10
```
After that negotioantion I see traffic (I think with video): from device  ip:192.168.15.111 port:49624 to iOs   ip:192.168.15.106 port:52142

**Comment by KonstantinBYvitebsk** [comment]
@stasel  Hi.  
I suppose the application is based on the framework https://github.com/stasel/WebRTC.
I can be wrong, but it seems to me the framework is not support the codec H264. Am I right?
If this package doesn't support the codec H264, can you advice me how to do it working?
I can't use another codec, so what I think (solutions to the problem).
1. Use webkit at the swift code.
2. Rebuild the framework after editing. I can try to enable H264 at code. I am not sure it is rightfully.
3. May be you can advice some other way?

**Comment by stasel** [comment]
I'm pretty sure that H264 support is build in WebRTC library. I just sent an offer from my iPhone X and saw H264 mentions in the SDP string.

Yes, I use the framework in https://github.com/stasel/WebRTC . You can try to compile WebRTC on your own - check the other repo for instructions and a handy script. Maybe there is a special flag or something.

Let me know how it goes

**Comment by KonstantinBYvitebsk** [comment]
Hi @stasel. Thanks for your answers.

I'll explain why I think it is not added:
1.  In my case it is not working (so-so explanation :) )
2. When I try to find text which that can be thrown away by definition  
```
#ifdef WEBRTC_USE_H264
#endif
```
I found that for example  **H264EncoderImpl** is not in the package.

`strings WebRTC |grep H264EncoderImpl
`
But we have function **SupportedH264Codecs**

`strings WebRTC |grep SupportedH264Codecs
`
And  yes. I also saw H264  when offers are created. But I see another interesting string at build scripts:

Also I see the strings at the file webrtc.gni.
```
# Enable this to build OpenH264 encoder/FFmpeg decoder. This is supported on
# all platforms except Android and iOS. Because FFmpeg can be built
# with/without H.264 support, |ffmpeg_branding| has to separately be set to a
# value that includes H.264, for example "Chrome". If FFmpeg is built without
# H.264, compilation succeeds but |H264DecoderImpl| fails to initialize.
# CHECK THE OPENH264, FFMPEG AND H.264 LICENSES/PATENTS BEFORE BUILDING.
# http://www.openh264.org, https://www.ffmpeg.org/
#
# Enabling H264 when building with MSVC is currently not supported, see
# bugs.webrtc.org/9213#c13 for more info.
rtc_use_h264 =
    proprietary_codecs && !is_android && !is_ios && !(is_win && !is_clang)

```
 I am still finding roots of the problem.

**Reply by stasel** [reply]
Great research!
So H264 is not included in iOS or Android builds due to licensing reasons.
I'm trying now to compile the library with `rtc_use_h264=true` and `proprietary_codecs=true` 
I'll report here how it goes

Edit: Unfortunately build fails when I added the following flags: `rtc_use_h264=true proprietary_codecs=true ffmpeg_branding="Chrome"`

I get an error:
```
In file included from ../../third_party/ffmpeg/libavcodec/aacps_common.c:23:
../../third_party/ffmpeg/libavutil/common.h:45:10: fatal error: 'libavutil/avconfig.h' file not found
#include "libavutil/avconfig.h"
         ^~~~~~~~~~~~~~~~~~~~~~
1 error generated.
[1850/3675] CC obj/third_party/boringssl/boringssl/bcm.o
ninja: build stopped: subcommand failed.
```

A further investigation is required. Maybe you can check if other people managed to compile iOS version with H264 codecs enabled

**Reply by KonstantinBYvitebsk** [reply]
So. libavutil is part of the ffmpeg
https://ffmpeg.org/libavutil.html

Unfortunately I don't have macbook to compile the framework myself.

**Reply by stasel** [reply]
I did manage to compile a macOS version of the library with H264 enabled using the flags above. In the docs it says the H264 is not supported for iOS and Android. 
I don't really know how to continue from here. Maybe it's worth for you to ask around WebRTC developers: https://groups.google.com/g/discuss-webrtc

**Reply by KonstantinBYvitebsk** [reply]
Cool. 
I've created the discussion in the group already.
Now I am waiting for answers. I've provided a couple of link  to yours repo. So I am going to add link to the discussion there after the link to discussion appears.

**Reply by KonstantinBYvitebsk** [reply]
Hi @stasel 
It seems to me I confused myself and confused you.
I have managed to create web page (html+js) which can connect to your signal server. Even can show video and play sound.
It works  not stable for me(frozen sometimes, audio without video) but it has shown me that it uses H264. And it decode/encode with H264.
[ios-demo-html.tar.gz](https://github.com/stasel/WebRTC-iOS/files/7094878/ios-demo-html.tar.gz)
It was made on gst-examples base.  It works at least on chrome .
To connect:
1. Unpack archive.
2. Start signal server.
3. Open index.html in browser.
4. Enter server IP.
5. Press "connect"
6. Connect DEMO to the signal server.
7. Send offer.
